app/cdap/components/SourceControlManagement/store/ActionCreator.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { MyPipelineApi } from 'api/pipeline';
+import { BATCH_PIPELINE_TYPE } from 'services/helpers';
+import SourceControlManagementSyncStore, { PushToGitActions } from '.';
+import debounce from 'lodash/debounce';
+import { MyRepositoryApi } from 'api/repository';
+import { catchError, concatMap, map } from 'rxjs/operators';
+import { of } from 'rxjs/observable/of';
+import { ILocalPipeline } from '../types';
+import T from 'i18n-react';
+
+export const getNamespacePipelineList = (namespace, nameFilter = null) => {
+  MyPipelineApi.list({
+    namespace,
+    artifactName: BATCH_PIPELINE_TYPE,
+    nameFilter,
+  }).subscribe((res) => {
+    let pipelines = Array.isArray(res) ? res : res?.applications;
+    pipelines = pipelines.map((pipeline) => {
+      return {
+        name: pipeline.name,
+        fileHash: pipeline.sourceControlMeta?.fileHash,
+        error: null,
+        success: null,
+      };
+    });
+    setLocalPipelines(pipelines);
+  });
+};
+
+export const setLocalPipelines = (pipelines: ILocalPipeline[]) => {
+  SourceControlManagementSyncStore.dispatch({
+    type: PushToGitActions.setLocalPipelines,
+    payload: {
+      localPipelines: pipelines,
+    },
+  });
+};
+
+const applySearch = () => {
+  SourceControlManagementSyncStore.dispatch({
+    type: PushToGitActions.applySearch,
+  });
+};
+const debouncedApplySearch = debounce(applySearch, 1000);
+
+export const setNameFilter = (nameFilter: string) => {
+  SourceControlManagementSyncStore.dispatch({
+    type: PushToGitActions.setNameFilter,
+    payload: {
+      nameFilter,
+    },
+  });
+  debouncedApplySearch();
+};
+
+export const setSelectedPipelines = (selectedPipelines: any[]) => {
+  SourceControlManagementSyncStore.dispatch({
+    type: PushToGitActions.setSelectedPipelines,
+    payload: {
+      selectedPipelines,
+    },
+  });
+};
+
+export const pushSelectedPipelines = (namespace, apps, payload, loadingMessageDispatcher) => {
+  return of(...apps).pipe(
+    concatMap((appId) => {
+      loadingMessageDispatcher(pushAppMessage(appId));
+      const params = {
+        namespace,
+        appId,
+      };
+      return MyRepositoryApi.push(params, payload).pipe(
+        map((res: object) => ({ ...res, statusCode: 200 })),
+        catchError((err) => {
+          return of({ ...err, name: appId });
+        })
```

Comment:
@njbriggs can you check the design of this api call? I want to send synchronous request of each `app` in `apps`, so I'm using `concatMap` to achieve it. 
And I also want all requests to finish whether any one of them failed, so in `catchError` I return the `err` as an observable for further processing. Then in the caller, both the success response and error will be in `next`, and I'm using a `statusCode` to differentiate whether it's success or failure.

Reply:
https://rxjs.dev/operator-decision-tree

You could try `forkJoin`. Let me see if I can get this right:

```
forkJoin(apps.map(app => {
  ...
  return MyRepositoryApi.push(params, payload);
})
```

However:

```
If any given observable errors at some point, [forkJoin](https://rxjs.dev/api/index/function/forkJoin) will error as well and immediately unsubscribe from the other observables.
```

which is not the behavior you want. So you could keep the `pipe` on each inner Observable to catch the error and convert resolve object that can differentiate between success and error.